drivers/flash/nrf_qspi_nor.c

```diff
@@ -0,0 +1,690 @@
+	if ((addr < 0) || ((addr + size) > params->size)) {
+		return -EINVAL;
+	}
+	struct qspi_buf rx_buf = {
+		.buf = dest,
+		.len = size
+	};
+
+	int ret = nrfx_qspi_read(rx_buf.buf, rx_buf.len, addr);
+
+	qspi_wait_for_completion(dev);
+	return qspi_get_zephyr_ret_code(ret);
+}
+
+static int qspi_nor_write(struct device *dev, off_t addr, const void *src,
+			  size_t size)
+{
+	if (!src) {
+		return -EINVAL;
+	}
+
+	/* write size must be multiple of 4 bytes */
+	if (size % sizeof(uint32_t) ||
+	    !(size > 0)) {
+		return -EINVAL;
+	}
+
+	qspi_lock(dev);
+
+	struct qspi_nor_data *const driver_data = dev->driver_data;
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	if (driver_data->write_protection) {
+		return -EACCES;
+	}
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((size + addr) > params->size)) {
+		return -EINVAL;
+	}
+	struct qspi_buf tx_buf = {
+		.buf = (void *)src,
+		.len = size
+	};
+
+	int ret = nrfx_qspi_write(tx_buf.buf, tx_buf.len, addr);
+
+	qspi_wait_for_completion(dev);
+	return qspi_get_zephyr_ret_code(ret);
+}
+
+static int qspi_nor_erase(struct device *dev, off_t addr, size_t size)
+{
+	struct qspi_nor_data *const driver_data = dev->driver_data;
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	if (driver_data->write_protection) {
+		return -EACCES;
+	}
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((size + addr) > params->size)) {
+		return -ENODEV;
+	}
+
+	int ret = qspi_erase(dev, addr, size);
+
+	return ret;
+}
+
+static int qspi_nor_write_protection_set(struct device *dev,
+					 bool write_protect)
+{
+	struct qspi_nor_data *const driver_data = dev->driver_data;
+
+	int ret = 0;
+	struct qspi_cmd cmd = {
+		.op_code = ((write_protect) ? SPI_NOR_CMD_WRDI : SPI_NOR_CMD_WREN),
+	};
+
+	driver_data->write_protection = write_protect;
+
+	if (qspi_send_cmd(dev, &cmd) != 0) {
+		ret = -EIO;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Configure the flash
+ *
+ * @param dev The flash device structure
+ * @param info The flash info structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static int qspi_nor_configure(struct device *dev)
+{
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	int ret = qspi_nrfx_configure(dev);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* now the spi bus is configured, we can verify the flash id */
+	if (qspi_nor_read_id(dev, params) != 0) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Initialize and configure the flash
+ *
+ * @param name The flash name
+ * @return 0 on success, negative errno code otherwise
+ */
+static int qspi_nor_init(struct device *dev)
+{
+	IRQ_CONNECT(DT_NORDIC_NRF_QSPI_QSPI_0_IRQ_0,
+		    DT_NORDIC_NRF_QSPI_QSPI_0_IRQ_0_PRIORITY,
+		    nrfx_isr, nrfx_qspi_irq_handler, 0);
+	return qspi_nor_configure(dev);
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+
+/* instance 0 size in bytes */
+#define INST_0_BYTES (DT_INST_0_NORDIC_QSPI_NOR_SIZE / 8)
```

Comment:
move INST_0_BYTES outside of ifdefry in case using in device initialization macro